### PR TITLE
REL-8142 ServiceNow blueprint

### PR DESCRIPTION
**Pull request by RoelBoitenX**
## Definition of Done

This is the definition of done for (new and modified) blueprints to be accepted into the [central XebiaLabs blueprints repository](https://github.com/xebialabs/blueprints).

### Value

This section describes how to determine whether a blueprint has enough value to be included in the central XebiaLabs blueprints repository.

- [X] The blueprint applies to a focused use case.
- [X] The blueprint uses at least one XebiaLabs product.
- [X] The blueprint depends on as few other products as possible. **N.B.:** This is part of the definition of done to keep the blueprint focused and applicable to as many users as possible. A blueprint that depends on one specific CI tool **and** one specific Java EE application server **and** one specific ticketing system can only be used by users that have all three products.
- [X] The blueprint can be used with content supplied by the user but also includes demo content. A blueprint that does not work with content supplied by the user is a demo or a showcase and does not help the user with their own work.

### Technical

This section describes how to determine whether the blueprint has the right technical quality to be included in the central XebiaLabs blueprints repository.

- [X] The branch from which the PR is created is up-to-date with the `development` branch.
- [X] The blueprint contains a `README.md` file at the root of the blueprint folder that describes the blueprint, using the README template in the `.github` folder of this repository.
- [X] The blueprint generates a `xebialabs/USAGE.md` file which, at a minimum, explains how to use the blueprint after it is generated (like adding any missing steps, creating accounts, setting up Docker containers, applying the YAML using `xl` CLI, running release. etc.). **N.B.:** Do not use this document to describe how to instantiate the blueprint. It will only be available to the user *after* the blueprint has been instantiated.
- [x] The blueprint generates a `docker/docker-compose.yml` file which can be used to spin up the required products that the blueprints uses. This can be used to test and/or try out the blueprint without having to manually install those products.
- [X] The blueprint does not contain sensitive information such passwords, tokens, credentials or licenses.
- [X] The blueprint uses `secret: true` in the `blueprint.yaml` parameter definition for question that ask for sensitive information.
- [X] The blueprint does not contain the files `xebialabs/.gitignore`, `xebialabs/values.xlvals` and `xebialabs/secrets.xlvals` and does not refer to them from the files section of the `blueprint.yaml` file. These files will be generated when the blueprint is instantiated. To generate the `xebialabs/values.xlvals` file, use the `saveInXlVals: true` directive in the parameters section of the blueprint. To generate the `xebialabs/secrets.xlvals` file, use the `secret: true` directive .
- [X] The blueprint does not define parameters for trivial things like phase names, task names, folder names etc. Ask questions that add value and be opinionated where possible. For example, ask for a project name and derive folder names, task names etc from that.
- [X] Folder and file names must use [kebab-case](http://wiki.c2.com/?KebabCase), for example: `aws/sample-app-demo`, `xld-environment.yaml`

### Review and testing

This section describes how to determine whether the blueprint has been reviewed and tested well enough to be included in the central XebiaLabs blueprints repository.

- [X] Unit test is added in a `__test__` folder for each blueprint. Refer the `CONTRIBUTING.md` file at the root of of this repository for more details.
- [x] The Travis CI for the PR is green
- [ ] The blueprint has been reviewed by someone else in the team.
- [ ] Both README and USAGE files have been reviewed by a technical writer and a product marketing manager.
- [ ] The blueprint has been manually tested with the `docker/docker-compose.yml` that is generated as part of it.
- [ ] The blueprint works on Linux, Mac and Windows.

### CODE REVIEW

**Comment by sffortytwo** [comment]
Hi Roel. Use this docker directory with its docker-compose.yml to satisfy the one tick box.

**Comment by sffortytwo** [comment]
The four lines reading useServicenowApp: true|false in the two YAML files are indented 2 spaces. This breaks the YAML formatting. Please unindent them

File: servicenow/sample-templates/blueprint.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    description: What is the user name of the ServiceNow user?
+  - name: ServiceNowPassword
+    type: Input
+    secret: true
+    description: What is the password of the ServiceNow user?
+  - name: ServiceNowApp
+    type: Confirm
+    description: Will you be using the ServiceNow app?
+
+  files:
```
Comment (jkrieger-xl) [inline servicenow/sample-templates/blueprint.yaml]: It doesn't include the docker-compose file

File: servicenow/sample-templates/xebialabs/ITSM_sample.yaml.tmpl [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+  - template: ITSM from ServiceNow
+    description: ITSM started from a change request in ServiceNow;
+    phases:
+    - phase: Acceptance
+      tasks:
+      - name: Inform Servicenow start Dev
+        type: servicenow.UpdateChangeRequest
+        servicenowServer: {{.ServerName}}
+        sysId: ${id}
+      - name: Deploy to TEST (XL Deploy)
```
Comment (jkrieger-xl) [inline servicenow/sample-templates/xebialabs/ITSM_sample.yaml.tmpl]: There is no xl deploy container

File: servicenow/sample-templates/blueprint.yaml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    The blueprint supplies two yaml files with sample templates for usage with the XL Release ServiceNow plugin and optionnaly the ServiceNow application 'XL Release Integration' from teh appstore.
+  author: XebiaLabs
+  version: 1.0
+  instructions: Please read the generated file "xebialabs/USAGE.md" for further usage instructions.
+spec:
+  parameters:
+  - name: ServerName
+    type: Input
+    description: Provide a name for the ServiceNow server?
+  - name: ServiceNowURL
```
Comment (jkrieger-xl) [inline servicenow/sample-templates/blueprint.yaml]: is it possible to provide a ServiceNow demo container image?

File: servicenow/sample-templates/xebialabs/agile_ITSM_sample.yaml.tmpl [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+spec:
+- directory: ServiceNow
+  children:
+  - name: {{.ServerName}}
+    type: servicenow.Server
+    url: {{.ServiceNowURL}}
+    username: {{.ServiceNowUser}}
+    password: {{.ServiceNowPassword}}
+    {{if .ServiceNowApp}}
+      useServicenowApp: true
```
Comment (jkrieger-xl) [inline servicenow/sample-templates/xebialabs/agile_ITSM_sample.yaml.tmpl]: wrong indentation

File: servicenow/sample-templates/xebialabs/ITSM_sample.yaml.tmpl [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+spec:
+- directory: ServiceNow
+  children:
+  - name: {{.ServerName}}
+    type: servicenow.Server
+    url: {{.ServiceNowURL}}
+    username: {{.ServiceNowUser}}
+    password: {{.ServiceNowPassword}}
+    {{if .ServiceNowApp}}
+      useServicenowApp: true
```
Comment (jkrieger-xl) [inline servicenow/sample-templates/xebialabs/ITSM_sample.yaml.tmpl]: wrong indentation

File: servicenow/sample-templates/xebialabs/USAGE.md.tmpl [hunk 6]
```diff
@@ -0,0 +1,19 @@
+The blueprint supplies two yaml files with sample templates for usage with the ServiceNow application from the appstore.
+
+To deploy this blueprint with the XebiaLabs DevOps Platform, follow the steps below:
```
Comment (jkrieger-xl) [inline servicenow/sample-templates/xebialabs/USAGE.md.tmpl]: Missing docker-compose part and also please mention that the user needs to mount a volume with the ServiceNow plugin in it

Comment (jkrieger-xl) [reply servicenow/sample-templates/xebialabs/USAGE.md.tmpl]: It's explained here https://docs.xebialabs.com/xl-deploy/concept/docker-images-for-xl-deploy.html#plugins-directory-plugins

**Comment by sffortytwo** [comment]
Seems the ServiceNow plugin is not part of the base image. It will need to be added. Please make a note of this

File: servicenow/sample-templates/xebialabs/ITSM_sample.yaml.tmpl (hunk 2 above)
Comment (RoelBoitenX) [reply servicenow/sample-templates/xebialabs/ITSM_sample.yaml.tmpl]: Removed XL deploy dependency

File: servicenow/sample-templates/blueprint.yaml (hunk 3 above)
Comment (RoelBoitenX) [reply servicenow/sample-templates/blueprint.yaml]: Not a recent version that is supported by ServicNow. I have zero docker experenice, so can't say if creating one is even an option.

File: servicenow/sample-templates/xebialabs/USAGE.md.tmpl (hunk 6 above)
Comment (RoelBoitenX) [reply servicenow/sample-templates/xebialabs/USAGE.md.tmpl]: Added a section that informs the user about the plugib and how to do that.

File: servicenow/sample-templates/blueprint.yaml (hunk 3 above)
Comment (jkrieger-xl) [reply servicenow/sample-templates/blueprint.yaml]: How can we test this blueprint? Can you provide us some guidelines for someone that has zero ServiceNow experience?

**Pull request closed without merging**